### ISSUE TITLE
Pydantic can't parse literal values from json via `(model_)validate_json` method

### ISSUE THREAD

**Issue by ThirVondukr** [pydantic/pydantic#8133]
### Initial Checks

- [X] I confirm that I'm using Pydantic V2

### Description

Pydantic fails when trying to parse literal values from dict/mapping in json

### Example Code

```Python
from typing import Literal

from pydantic import TypeAdapter

literal = Literal["a", "b"]


dict_adapter = TypeAdapter(dict[literal, int])
dict_adapter.validate_python({"a": 42})  # Ok
dict_adapter.validate_json('{"a": 42}')  # Raises an error
```

### Python, Pydantic & OS Version

```
$ pdm show pydantic
Name:                  pydantic
Latest version:        2.5.0
Latest stable version: 2.5.0
Installed version:     2.5.0
```
```
$ pdm show pydantic-core
Name:                  pydantic_core                                      
Latest version:        2.14.3
Latest stable version: 2.14.3
Installed version:     2.14.1
```

**Comment by Viicos** [comment]
`{42: "a"}` is not valid JSON (keys can only be strings).

But I'm able to reproduce the second example (this is a regression from 2.4.2)

**Comment by ThirVondukr** [comment]
@Viicos My bad, I thought it fails with literal values too, but that only seems to be the case for keys, I adjusted the MRE

**Comment by sydney-runkle** [comment]
@ThirVondukr,

I can confirm this is a bug introduced by 2.5 as well. We'll work on a fix for this for our `2.5.2` patch release which we'll get out ASAP!

**Comment by sydney-runkle** [comment]
Closing this as we've merged the related PR in `pydantic-core`.
